### DISCUSSION TITLE
CircuitPython for ESP32S3 | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#565] [Q&A] [answered]
# CircuitPython for ESP32S3 | Seeed Studio Wiki

CircuitPython for ESP32S3

https://wiki.seeedstudio.com/XIAO_ESP32S3_CircuitPython/

**Comment by Parkre01** [comment]
This was a great tutorial.  I just got my SEEED ESP32-S3.  I was able to flash it to: 

Espressif - ESP32-S3-DevKitM-1-N8  like you did.  I did not get any errors.  However, I did not get the "CIRCUITPY" on the USB like you did. I tried running a quick LED blink test to verify it worked, but kept getting an error. It kept indicating that I had no libraries installed for Circuit Python. 


++++++++++++++++++++++++++++++++
import board
import digitalio
import time

led_pin = digitalio.DigitalInOut(board.D2)  # GPIO pin connected to the LED (i tried external led too, D10 to ground, failed)
led_pin.direction = digitalio.Direction.OUTPUT

delay_time = 1  # Delay time in seconds

while True:
    led_pin.value = True  # Turn the LED on
    time.sleep(delay_time)  # Wait for the specified delay time
    led_pin.value = False  # Turn the LED off
    time.sleep(delay_time)  # Wait for the specified delay time

++++++++++++++++++++++++++++++++++++++++++++

In thonny, I have  lib folder, but there's nothing in it. 

Does this Seeed ESP32-S3 actually support CircuitPython?

**Reply by xzxcessarr** [reply, marked as answer]
Hello, our board is support the circuitpython, the reason you cannot use is the name of the pin is error, please read the pin picture here: https://wiki.seeedstudio.com/xiao_esp32s3_getting_started/#hardware-overview, if you wanna use the D2 pin, you will see it matchs the GPIO3,  so the name of the pin is the "IO3", you can use the board.IO3 to call D2 pin

**Reply by Parkre01** [reply]
Thanks!!!! I got my external flashing LED to work!


++++++++++++++++++++++++++++++++++++
import board
import digitalio
import time


print("hello, world")
# Define the GPIO pin for the LED
led_pin = digitalio.DigitalInOut(board.IO9)  # 
led_pin.direction = digitalio.Direction.OUTPUT

# Variable to track the LED state
led_state = False

# Infinite loop to flash the LED
while True:
    led_state = not led_state  # Toggle the LED state (on to off or off to on)
    led_pin.value = led_state  # Set the LED state
    print(led_state)
    time.sleep(1)             # Wait for 1 second


+++++++++++++++++++++++++++++++++++++++++++++++

I'm currently got CircuitPython 8.2.4 installed. 

Below is populated when the ESP-32-S3 is turned on. 
]0;🐍Wi-Fi: off | REPL | 8.2.4\
Adafruit CircuitPython 8.2.4 on 2023-08-22; ESP32-S3-DevKitC-1-N8 with ESP32S3

**Comment by layogtima** [comment]
Love the XIAO ESP32C3! 

I've manged to make WiFi work flawlessly over Circuitpython. 

Has been anyone been able to get BLE over CircuitPython working?

https://github.com/adafruit/circuitpython/issues/7180

Found this as a dead-end for the moment 💀

**Comment by kornpow** [comment]
I was able to get around using Thonny by using ESPtool directly.

I followed mostly the instructions related to micropython here: https://wiki.seeedstudio.com/XIAO_ESP32S3_Micropython/

Download circuitpython from here: https://circuitpython.org/board/espressif_esp32s3_devkitm_1_n8/
```
esptool.py \
	--port /dev/tty.usbmodem114301 --baud 460800 \
	--before default_reset --after hard_reset \
	--chip esp32s3  write_flash \
	--flash_mode dio --flash_size detect \
	--flash_freq 80m 0x0 ~/Downloads/adafruit-circuitpython-espressif_esp32s3_devkitm_1_n8-en_US-9.0.0.bin
```

**Comment by djairjr** [comment]
I'm compiling a new Circuitpython firmware for this board, following the information from the xlsx file available on Wiki: https://files.seeedstudio.com/wiki/SeeedStudio-XIAO-ESP32S3/res/XIAO_ESP32S3_Sense_Pinout.xlsx

I'm compiling a new Circuitpython firmware for this board, following the information from the xlsx file available on Wiki: https://files.seeedstudio.com/wiki/SeeedStudio-XIAO-ESP32S3/res/XIAO_ESP32S3_Sense_Pinout.xlsx 

I've made the following changes to the mpconfigboard.h, mpconfigboard.mk and pins.c files, to match the pinouts of the other cards in the XIAO series.

mpconfigboard.h file:

```
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

// Micropython setup

#define MICROPY_HW_BOARD_NAME       "Seeed Xiao ESP32-S3 Sense"
#define MICROPY_HW_MCU_NAME         "ESP32S3"

#define DEFAULT_UART_BUS_RX         (&pin_GPIO44)
#define DEFAULT_UART_BUS_TX         (&pin_GPIO43)

#define DEFAULT_SPI_BUS_SCK (&pin_GPIO7)
#define DEFAULT_SPI_BUS_MOSI (&pin_GPIO9)
#define DEFAULT_SPI_BUS_MISO (&pin_GPIO8)

#define DEFAULT_I2C_BUS_SCL (&pin_GPIO6)
#define DEFAULT_I2C_BUS_SDA (&pin_GPIO5)
```

mpconfigboard.mk file:

```
USB_VID = 0x303A
USB_PID = 0x1001
USB_PRODUCT = "Seeed Xiao ESP32-S3 Sense"
USB_MANUFACTURER = "Seeed Studio"

IDF_TARGET = esp32s3

CIRCUITPY_ESP_FLASH_MODE = qio
CIRCUITPY_ESP_FLASH_FREQ = 80m
CIRCUITPY_ESP_FLASH_SIZE = 8MB
CIRCUITPY_ESPCAMERA = 1
```


and finally  pins.c file:

```
#include "shared-bindings/board/__init__.h"

STATIC const mp_rom_map_elem_t board_module_globals_table[] = {
    CIRCUITPYTHON_BOARD_DICT_STANDARD_ITEMS

    { MP_ROM_QSTR(MP_QSTR_A0), MP_ROM_PTR(&pin_GPIO1) },
    { MP_ROM_QSTR(MP_QSTR_D0), MP_ROM_PTR(&pin_GPIO1) },
    { MP_ROM_QSTR(MP_QSTR_A1), MP_ROM_PTR(&pin_GPIO2) },
    { MP_ROM_QSTR(MP_QSTR_D1), MP_ROM_PTR(&pin_GPIO2) },
    { MP_ROM_QSTR(MP_QSTR_A2), MP_ROM_PTR(&pin_GPIO3) },
    { MP_ROM_QSTR(MP_QSTR_D2), MP_ROM_PTR(&pin_GPIO3) },
    { MP_ROM_QSTR(MP_QSTR_A3), MP_ROM_PTR(&pin_GPIO4) },
    { MP_ROM_QSTR(MP_QSTR_D3), MP_ROM_PTR(&pin_GPIO4) },    
    { MP_ROM_QSTR(MP_QSTR_A4), MP_ROM_PTR(&pin_GPIO5) },
    { MP_ROM_QSTR(MP_QSTR_SDA), MP_ROM_PTR(&pin_GPIO5) },
    { MP_ROM_QSTR(MP_QSTR_D4), MP_ROM_PTR(&pin_GPIO5) },    
    { MP_ROM_QSTR(MP_QSTR_A5), MP_ROM_PTR(&pin_GPIO6) },
    { MP_ROM_QSTR(MP_QSTR_SCL), MP_ROM_PTR(&pin_GPIO6) },
    { MP_ROM_QSTR(MP_QSTR_D5), MP_ROM_PTR(&pin_GPIO6) },   
    { MP_ROM_QSTR(MP_QSTR_D6), MP_ROM_PTR(&pin_GPIO43) },
    { MP_ROM_QSTR(MP_QSTR_TX), MP_ROM_PTR(&pin_GPIO43) },       
    { MP_ROM_QSTR(MP_QSTR_D7), MP_ROM_PTR(&pin_GPIO47) },
    { MP_ROM_QSTR(MP_QSTR_RX), MP_ROM_PTR(&pin_GPIO47) },
    { MP_ROM_QSTR(MP_QSTR_D8), MP_ROM_PTR(&pin_GPIO7) },    
    { MP_ROM_QSTR(MP_QSTR_D9), MP_ROM_PTR(&pin_GPIO8) },
    { MP_ROM_QSTR(MP_QSTR_D10), MP_ROM_PTR(&pin_GPIO9) },
    
    { MP_ROM_QSTR(MP_QSTR_SCK), MP_ROM_PTR(&pin_GPIO7) },    
    { MP_ROM_QSTR(MP_QSTR_MISO), MP_ROM_PTR(&pin_GPIO8) },
    { MP_ROM_QSTR(MP_QSTR_MOSI), MP_ROM_PTR(&pin_GPIO9) },
    
    { MP_ROM_QSTR(MP_QSTR_XCLK), MP_ROM_PTR(&pin_GPIO10) },
    { MP_ROM_QSTR(MP_QSTR_Y8), MP_ROM_PTR(&pin_GPIO11) },
    { MP_ROM_QSTR(MP_QSTR_Y7), MP_ROM_PTR(&pin_GPIO12) },
    { MP_ROM_QSTR(MP_QSTR_PCLK), MP_ROM_PTR(&pin_GPIO13) },
    { MP_ROM_QSTR(MP_QSTR_Y6), MP_ROM_PTR(&pin_GPIO14) },
    { MP_ROM_QSTR(MP_QSTR_Y2), MP_ROM_PTR(&pin_GPIO15) },
    { MP_ROM_QSTR(MP_QSTR_Y5), MP_ROM_PTR(&pin_GPIO16) },
    { MP_ROM_QSTR(MP_QSTR_Y3), MP_ROM_PTR(&pin_GPIO17) },
    { MP_ROM_QSTR(MP_QSTR_Y4), MP_ROM_PTR(&pin_GPIO18) },
    
    { MP_ROM_QSTR(MP_QSTR_LED), MP_ROM_PTR(&pin_GPIO21) },
    { MP_ROM_QSTR(MP_QSTR_SDCS), MP_ROM_PTR(&pin_GPIO21) },
 
    { MP_ROM_QSTR(MP_QSTR_VSYNC), MP_ROM_PTR(&pin_GPIO38) },
    { MP_ROM_QSTR(MP_QSTR_MTCK), MP_ROM_PTR(&pin_GPIO39) },
    { MP_ROM_QSTR(MP_QSTR_MTDO), MP_ROM_PTR(&pin_GPIO40) },
      
    { MP_ROM_QSTR(MP_QSTR_I2S_MIC), MP_ROM_PTR(&pin_GPIO41) },
    { MP_ROM_QSTR(MP_QSTR_MTDI), MP_ROM_PTR(&pin_GPIO41) },    
    { MP_ROM_QSTR(MP_QSTR_MTMS), MP_ROM_PTR(&pin_GPIO42) }, 
    { MP_ROM_QSTR(MP_QSTR_I2S_FS), MP_ROM_PTR(&pin_GPIO42) },
    
    { MP_ROM_QSTR(MP_QSTR_UART), MP_ROM_PTR(&board_uart_obj) },
};
MP_DEFINE_CONST_DICT(board_module_globals, board_module_globals_table);
```

I am compiling this version to see if it works, and I keep you informed...
